File: exercises/03_unknown_repository/scripts/processing.py
def call_variants(reference, alignments):
    """
    Identifies variants by comparing aligned reads to the reference.
    
    Args:
        reference (str): The reference DNA sequence.
        alignments (list): List of tuples containing (read_sequence, starting_position).
        
    Returns:
        dict: A dictionary mapping positions to found variants (ref_base, alt_base).
    """
    variants = {}
    for read_seq, pos in alignments:
        if pos == -1:
            continue
            
        for i, base in enumerate(read_seq):
            ref_pos = pos + i
            if ref_pos < len(reference):
                ref_base = reference[ref_pos]
                if base != ref_base:
                    if ref_pos not in variants:
                        variants[ref_pos] = []
                    variants[ref_pos].append({'ref': ref_base, 'alt': base})
                    
    return variants

File: exercises/03_unknown_repository_R/scripts/processing.R
call_variants <- function(reference, alignments) {
  # Identifies variants by comparing aligned reads to the reference.
  # 
  # Args:
  #   reference (character): The reference DNA sequence.
  #   alignments (list): List of lists containing 'read_sequence' and 'starting_position'.
  # 
  # Returns:
  #   list: A list mapping positions to found variants (ref_base, alt_base).
  
  variants <- list()
  
  for (align in alignments) {
    read_seq <- align$read_sequence
    pos <- align$starting_position
    
    if (pos == -1) {
      next
    }
    
    bases <- strsplit(read_seq, "")[[1]]
    ref_bases <- strsplit(reference, "")[[1]]
    
    for (i in seq_along(bases)) {
      base <- bases[i]
      ref_pos <- pos + (i - 1) + 1
      if (ref_pos <= length(ref_bases)) {
        ref_base <- ref_bases[ref_pos]
        if (base != ref_base) {
          pos_str <- as.character(ref_pos)
          if (is.null(variants[[pos_str]])) {
            variants[[pos_str]] <- list()
          }
          variants[[pos_str]] <- append(variants[[pos_str]], list(list(ref = ref_base, alt = base)))
        }
      }
    }
  }
  
  return(variants)
}
